File: LeetCode root folder/Minimum Path Cost in a Grid.cpp
//dp solution:0
//ms
//Beats
//100.00%

class Solution {
public:
	int minPathCost(vector<vector<int>>& grid, vector<vector<int>>& moveCost) {
		int m = grid.size();
		int n = grid[0].size();
		vector<int> dp = grid[m-1];
		for (int i = m-2; i >= 0; i--) {
			vector<int> tmp;
			for (int j = 0; j < n; j++) {
				int minCost = moveCost[grid[i][j]][0] + dp[0];
				for (int k = 0; k < n; k++) {
					int curCost = moveCost[grid[i][j]][k] + dp[k];
					minCost = min(minCost, curCost);
				}
				tmp.push_back(minCost + grid[i][j]);
			}
			dp = tmp;
		}
		return *min_element(dp.begin(), dp.end());
	}
};

File: LeetCode root folder/Minimum Path Cost in a Grid.py
# dp solution:167
# ms
# Beats
# 80.93%
class Solution:
    def minPathCost(self, grid: List[List[int]], moveCost: List[List[int]]) -> int:
        dp = grid[-1]
        for i in range(len(grid) - 2, -1, -1):
            tmp = []
            for j in range(len(grid[i])):
                minCost = float('inf')
                for k, cost in enumerate(moveCost[grid[i][j]]):
                    curCost = cost + dp[k]
                    minCost = min(minCost, curCost)
                tmp.append(minCost + grid[i][j])
            dp = tmp
        return min(dp)

